File: harness/javascript/runner.js
#!/usr/bin/env node
/* Generic JavaScript test runner driven by spec.yaml + tests/cases/partN.yaml.
 *
 * Invoked from dashboard/server.js as:
 *     node runner.js --problem-dir <tmpdir> --part <N>
 *
 * Emits one `PASS <name>` or `FAIL <name>` per test case, then a
 * `PART<N>_SUMMARY <passed>/<total>` line. This output contract is identical
 * to the Python/Java/C++ runners, so server.js's parseTestOutput consumes it
 * unchanged.
 *
 * Mirrors harness/python/runner.py semantics:
 *   • functions may be a list of {name,...} or a dict keyed by name
 *   • args are deserialized from YAML into solution class instances
 *   • the same assertion vocabulary is supported (expect / expect_equals /
 *     expect_close / expect_size / expect_field / also / expect_throws)
 *
 * Convention (must match scripts/gen_stubs.py JS emitter):
 *   solution.js exports every top-level class and function via CommonJS,
 *   and data-class constructors take positional args in declared field order.
 */

'use strict';

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

// ─── arg parsing ──────────────────────────────────────────────────────────────

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--problem-dir') out.problemDir = argv[++i];
    else if (argv[i] === '--part') out.part = parseInt(argv[++i], 10);
  }
  if (!out.problemDir || !Number.isInteger(out.part)) {
    console.error('usage: node runner.js --problem-dir <dir> --part <N>');
    process.exit(2);
  }
  return out;
}

// ─── solution loading ───────────────────────────────────────────────────────

function loadSolution(problemDir) {
  const solPath = path.join(problemDir, 'solution.js');
  // require() caches by absolute path; each runner process is fresh so this is fine.
  const mod = require(solPath);
  if (!mod || typeof mod !== 'object') {
    throw new Error('solution.js must export its classes and functions (module.exports = { ... }).');
  }
  return mod;
}

// ─── spec helpers ───────────────────────────────────────────────────────────

// `functions` may be a list of {name,...} or a dict keyed by name.
function normalizeFunctions(spec) {
  const fns = spec.functions || {};
  if (Array.isArray(fns)) {
    const map = {};
    for (const f of fns) map[f.name] = f;
    return map;
  }
  return fns;
}

// Return the inner type for `list<X>`, else null.
function parseListType(t) {
  if (typeof t === 'string' && t.startsWith('list<') && t.endsWith('>')) {
    return t.slice(5, -1);
  }
  return null;
}

// ─── argument deserialization ───────────────────────────────────────────────

/* Instantiate a user-defined struct from a plain object.
 *
 * Fields are passed positionally in declared order. A field present in `value`
 * uses that value; a field absent but carrying a `default` in the spec falls
 * back to that default; an absent field with no default is a hard error. This
 * keeps construction order-stable and matches the positional ctor emitted by
 * gen_stubs.py for JS (and mirrors the C++/Java POJO convention). */
function deserializeStruct(value, typeName, types, factories, solution) {
  const typeDef = types[typeName];
  const fields = typeDef.fields || [];
  const Cls = solution[typeName];
  if (typeof Cls !== 'function') {
    throw new Error(`solution.js does not export class '${typeName}'`);
  }
  const args = [];
  for (const f of fields) {
    const fname = f.name;
    if (Object.prototype.hasOwnProperty.call(value, fname)) {
      args.push(value[fname]);
    } else if (Object.prototype.hasOwnProperty.call(f, 'default')) {
      args.push(f.default);
    } else {
      throw new Error(`missing required field '${fname}' for ${typeName}`);
    }
  }
  return new Cls(...args);
}

// Evaluate a factory expression (e.g. `new RewardsMaximizer()`) with every
// solution export in scope. JS factory exprs are syntactically identical to the
// Java/C++ ones, so we fall back through javascript → java → cpp.
function evalFactory(expr, solution) {
  const names = Object.keys(solution);
  const values = names.map(n => solution[n]);
  // eslint-disable-next-line no-new-func
  const fn = new Function(...names, `return (${expr});`);
  return fn(...values);
}

// Coerce a YAML-loaded value into the actual JS argument.
function deserializeArg(value, declaredType, types, factories, solution) {
  const inner = parseListType(declaredType);
  if (inner !== null) {
    if (!Array.isArray(value)) {
      throw new Error(`expected list for ${declaredType}, got ${typeof value}`);
    }
    return value.map(item => deserializeArg(item, inner, types, factories, solution));
  }

  if (declaredType === 'factory') {
    const entry = factories[value] || {};
    const expr = entry.javascript || entry.js || entry.java || entry.cpp;
    if (!expr) throw new Error(`factory '${value}' has no javascript/java/cpp expression`);
    return evalFactory(expr, solution);
  }

  if (types[declaredType] && value && typeof value === 'object' && !Array.isArray(value)) {
    return deserializeStruct(value, declaredType, types, factories, solution);
  }

  // Primitives (string/int/float/bool) and unknown types pass through.
  return value;
}

// ─── result inspection + comparison ──────────────────────────────────────────

function extractField(obj, field) {
  if (obj === null || obj === undefined) return undefined;
  return obj[field];
}

// Structural deep-equality matching Python's == for the value shapes that
// appear in test YAML (scalars, arrays, plain objects). Numbers compare with a
// tiny tolerance so YAML floats like 0.1 don't trip exact-equality.
function deepEqual(a, b) {
  if (a === b) return true;
  if (typeof a === 'number' && typeof b === 'number') {
    if (Number.isNaN(a) && Number.isNaN(b)) return true;
    return Math.abs(a - b) < 1e-9;
  }
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) if (!deepEqual(a[i], b[i])) return false;
    return true;
  }
  if (a && b && typeof a === 'object' && typeof b === 'object') {
    const ka = Object.keys(a), kb = Object.keys(b);
    if (ka.length !== kb.length) return false;
    for (const k of ka) {
      if (!Object.prototype.hasOwnProperty.call(b, k)) return false;
      if (!deepEqual(a[k], b[k])) return false;
    }
    return true;
  }
  return false;
}

// ─── per-test execution ───────────────────────────────────────────────────────

function runTest(testCase, fnSpecs, types, factories, solution) {
  const name = testCase.name;
  const fnName = testCase.call;
  const fnSpec = fnSpecs[fnName];
  if (!fnSpec) throw new Error(`spec has no function '${fnName}'`);
  const params = fnSpec.params || [];

  const rawArgs = testCase.args || [];
  const args = rawArgs.map((a, i) =>
    deserializeArg(a, params[i].type, types, factories, solution));

  const fn = solution[fnName];
  if (typeof fn !== 'function') {
    throw new Error(`solution.js does not export function '${fnName}'`);
  }

  const expectThrows = testCase.expect_throws || false;
  let result;
  try {
    result = fn(...args);
  } catch (e) {
    return [!!expectThrows, name];
  }
  if (expectThrows) return [false, name];

  if (Object.prototype.hasOwnProperty.call(testCase, 'expect_equals')) {
    if (!deepEqual(result, testCase.expect_equals)) return [false, name];
  }

  if (Object.prototype.hasOwnProperty.call(testCase, 'expect_close')) {
    const eps = testCase.epsilon !== undefined ? testCase.epsilon : 0.001;
    const r = Number(result), e = Number(testCase.expect_close);
    if (!(Math.abs(r - e) <= eps)) return [false, name];
  }

  if (Object.prototype.hasOwnProperty.call(testCase, 'expect_size')) {
    if (!result || result.length !== testCase.expect_size) return [false, name];
  }

  if (Object.prototype.hasOwnProperty.call(testCase, 'expect')) {
    const field = testCase.expect_field;
    const actual = field
      ? Array.from(result, item => extractField(item, field))
      : Array.from(result);
    if (!deepEqual(actual, testCase.expect)) return [false, name];
  }

  for (const check of testCase.also || []) {
    if (extractField(result[check.index], check.field) !== check.equals) {
      // fall back to deep compare for object/array equals
      if (!deepEqual(extractField(result[check.index], check.field), check.equals)) {
        return [false, name];
      }
    }
  }

  return [true, name];
}

// ─── main ─────────────────────────────────────────────────────────────────────

function main() {
  const { problemDir, part } = parseArgs(process.argv.slice(2));

  // Use { json: true } so duplicate mapping keys are tolerated (last value
  // wins) instead of throwing — matching PyYAML's safe_load leniency used by
  // the Python runner. Some specs carry a duplicated `types:` block; the two
  // runners must agree on how that's handled.
  const loadYaml = (file) => yaml.load(fs.readFileSync(file, 'utf8'), { json: true });
  const spec = loadYaml(path.join(problemDir, 'spec.yaml'));
  const casesPath = path.join(problemDir, 'tests', 'cases', `part${part}.yaml`);
  const cases = loadYaml(casesPath) || [];

  const solution = loadSolution(problemDir);
  const fnSpecs = normalizeFunctions(spec);
  const types = spec.types || {};
  const factories = spec.factories || {};

  let passed = 0;
  const total = cases.length;
  for (const testCase of cases) {
    let ok, name;
    try {
      [ok, name] = runTest(testCase, fnSpecs, types, factories, solution);
    } catch (e) {
      ok = false;
      name = (testCase && testCase.name) || '<unnamed>';
      process.stderr.write(String(e && e.stack ? e.stack : e) + '\n');
    }
    process.stdout.write(`${ok ? 'PASS' : 'FAIL'} ${name}\n`);
    if (ok) passed++;
  }

  process.stdout.write(`PART${part}_SUMMARY ${passed}/${total}\n`);
}

main();

File: harness/python/runner.py
#!/usr/bin/env python3
"""Generic Python test runner driven by spec.yaml + tests/cases/partN.yaml.

Invoked from dashboard/server.js as:
    python3 runner.py --problem-dir <tmpdir> --part <N>

Emits one `PASS <name>` or `FAIL <name>` per test case, then a
`PART<N>_SUMMARY <passed>/<total>` line.
"""

import argparse
import importlib.util
import os
import sys
import traceback

import yaml


def load_solution(problem_dir):
    sol_path = os.path.join(problem_dir, "solution.py")
    spec = importlib.util.spec_from_file_location("solution", sol_path)
    module = importlib.util.module_from_spec(spec)
    # Make the directory importable so any relative imports work.
    sys.path.insert(0, problem_dir)
    spec.loader.exec_module(module)
    return module


def normalize_functions(spec):
    """`functions` may be a list of {name,...} or a dict keyed by name."""
    fns = spec.get("functions", {})
    if isinstance(fns, list):
        return {f["name"]: f for f in fns}
    return fns


def parse_list_type(t):
    """Return the inner type for `list<X>`, else None."""
    if isinstance(t, str) and t.startswith("list<") and t.endswith(">"):
        return t[5:-1]
    return None


def deserialize_struct(value, type_name, types, factories, solution):
    """Instantiate a user-defined struct from a dict.

    Required fields (no `default` in spec) must be present in `value` and are
    passed positionally in declared order. Optional fields are passed as kwargs
    only when supplied — this lets callers omit any subset and rely on the
    class's default values, regardless of field ordering.
    """
    type_def = types[type_name]
    fields = type_def["fields"]
    cls = getattr(solution, type_name)
    pos_args = []
    kwargs = {}
    in_optional = False
    for f in fields:
        fname = f["name"]
        is_optional = "default" in f
        if not is_optional:
            if fname not in value:
                raise KeyError(f"missing required field '{fname}' for {type_name}")
            if in_optional:
                # Required fields after optional ones force kwarg-only mode.
                kwargs[fname] = value[fname]
            else:
                pos_args.append(value[fname])
        else:
            in_optional = True
            if fname in value:
                kwargs[fname] = value[fname]
    return cls(*pos_args, **kwargs)


def deserialize_arg(value, declared_type, types, factories, solution):
    """Coerce a YAML-loaded value into the actual Python argument."""
    inner = parse_list_type(declared_type)
    if inner is not None:
        if not isinstance(value, list):
            raise TypeError(f"expected list for {declared_type}, got {type(value)}")
        return [
            deserialize_arg(item, inner, types, factories, solution)
            for item in value
        ]

    if declared_type == "factory":
        # Factory key — eval the python expression in the solution namespace.
        expr = factories[value]["python"]
        return eval(expr, solution.__dict__)

    if declared_type in types and isinstance(value, dict):
        return deserialize_struct(value, declared_type, types, factories, solution)

    # Primitives (string/int/float/bool) and unknown types pass through.
    return value


def extract_field(obj, field):
    if isinstance(obj, dict):
        return obj.get(field)
    return getattr(obj, field)


def run_test(case, fn_specs, types, factories, solution):
    name = case["name"]
    fn_name = case["call"]
    fn_spec = fn_specs[fn_name]
    params = fn_spec["params"]

    raw_args = case.get("args", [])
    args = [
        deserialize_arg(raw_args[i], params[i]["type"], types, factories, solution)
        for i in range(len(raw_args))
    ]

    fn = getattr(solution, fn_name)

    expect_throws = case.get("expect_throws", False)
    try:
        result = fn(*args)
    except Exception:
        if expect_throws:
            return True, name
        return False, name

    if expect_throws:
        return False, name

    if "expect_equals" in case:
        if result != case["expect_equals"]:
            return False, name

    if "expect_close" in case:
        eps = case.get("epsilon", 0.001)
        try:
            if abs(float(result) - float(case["expect_close"])) > eps:
                return False, name
        except Exception:
            return False, name

    if "expect_size" in case:
        if len(result) != case["expect_size"]:
            return False, name

    if "expect" in case:
        field = case.get("expect_field")
        actual = (
            [extract_field(item, field) for item in result] if field else list(result)
        )
        if actual != case["expect"]:
            return False, name

    for check in case.get("also", []) or []:
        idx = check["index"]
        field = check["field"]
        expected = check["equals"]
        if extract_field(result[idx], field) != expected:
            return False, name

    return True, name


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--problem-dir", required=True)
    parser.add_argument("--part", required=True, type=int)
    args = parser.parse_args()

    with open(os.path.join(args.problem_dir, "spec.yaml")) as f:
        spec = yaml.safe_load(f)
    cases_path = os.path.join(
        args.problem_dir, "tests", "cases", f"part{args.part}.yaml"
    )
    with open(cases_path) as f:
        cases = yaml.safe_load(f) or []

    solution = load_solution(args.problem_dir)
    fn_specs = normalize_functions(spec)
    types = spec.get("types", {}) or {}
    factories = spec.get("factories", {}) or {}

    passed = 0
    total = len(cases)
    for case in cases:
        try:
            ok, name = run_test(case, fn_specs, types, factories, solution)
        except Exception:
            ok = False
            name = case.get("name", "<unnamed>")
            traceback.print_exc(file=sys.stderr)
        print(f"{'PASS' if ok else 'FAIL'} {name}")
        if ok:
            passed += 1

    print(f"PART{args.part}_SUMMARY {passed}/{total}")


if __name__ == "__main__":
    main()
